### PR TITLE
BUG: ArrowExtensionArray.to_numpy avoid object dtype when na_value provided

### LINKED ISSUE #54808: BUG: avoid conversion to object dtype in to_numpy for Arrow dtype and with dtype-compatible fill value
Using the current main branch (2.1.0rc0+115.gca429943c7), if we convert an Arrow-backed boolean array to numpy, we get object dtype if there are missing values:

```python
# no missing values -> bool dtype
>>> ser = pd.Series([True, False, True], dtype=pd.ArrowDtype(pa.bool_()))
>>> ser.to_numpy()
array([ True, False,  True])

# missing values -> object dtype
>>> ser = pd.Series([True, False, None], dtype=pd.ArrowDtype(pa.bool_()))
>>> ser.to_numpy()
array([True, False, <NA>], dtype=object)
```

This follows pyarrow's behaviour of converting to a numpy array (and I think also makes sense, unless we might want to raise an error instead, but that's for another discussion). 

But when a user specifies an `na_value` to avoid this issue of missing values with numpy, we still return an object dtype:

```python
>>> ser.to_numpy(na_value=False)
array([True, False, False], dtype=object)
```

while if the specified `na_value` is valid for the default target bool dtype (as in the example above), we could perfectly well return a proper bool dtype array automatically (currently you would have to be explicit: `to_numpy(np.bool_, na_value=False)`).

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] closes #54808
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.

### CODE REVIEW

File: pandas/core/arrays/arrow/array.py [hunk 1]
```diff
@@ -1240,46 +1240,50 @@ def to_numpy(
...
             if dtype is not None:
                 result = result.astype(dtype, copy=False)
             if copy:
                 result = result.copy()
-            return result
-        if self._hasna:
-            result[self.isna()] = na_value
+        else:
+            if dtype is None:
+                empty = data._pa_array[:0].to_numpy()
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:1278]: I think `data.dtype.empty._pa_array` would be a "nicer" way to get an empty example to use in `can_hold_element`

File: pandas/core/arrays/arrow/array.py [hunk 2]
```diff
@@ -1240,46 +1240,50 @@ def to_numpy(
...
-        elif pa.types.is_null(self._pa_array.type):
-            fill_value = None if isna(na_value) else na_value
-            return np.full(len(self), fill_value=fill_value, dtype=dtype)
-        elif self._hasna:
-            data = self.fillna(na_value)
+            result = np.array(list(data), dtype=dtype)
+            if data._hasna:
+                result[data.isna()] = na_value
+        elif pa.types.is_null(pa_type):
+            if dtype is not None and dtype.kind != "O" and isna(na_value):
```
Comment (mroeschke) [inline pandas/core/arrays/arrow/array.py:1267]: Why do we need to skip object here?

Comment (lukemanley) [reply pandas/core/arrays/arrow/array.py]: removed it, not needed

File: pandas/core/arrays/arrow/array.py (hunk 1 above)
Comment (lukemanley) [reply pandas/core/arrays/arrow/array.py]: yep, thanks. updated

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lukemanley

**Pull request merged**